email/exporter.go

```diff
@@ -0,0 +1,159 @@
+	berrors "github.com/letsencrypt/boulder/errors"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+// contactsQueueCap limits the queue size to prevent unbounded growth. This
+// value is adjustable as needed. Each RFC 5321 email address, encoded in UTF-8,
+// is at most 320 bytes. Storing 10,000 emails requires ~3.44 MB of memory.
+const contactsQueueCap = 10000
+
+var ErrQueueFull = errors.New("email-exporter queue is full")
+
+// ExporterImpl implements the gRPC server and processes email exports.
+type ExporterImpl struct {
+	emailpb.UnsafeExporterServer
+
+	sync.Mutex
+	drainWG sync.WaitGroup
+	wake    *sync.Cond
+
+	maxConcurrentRequests int
+	limiter               *rate.Limiter
+	toSend                []string
+	client                PardotClient
+	emailsHandledCounter  prometheus.Counter
+	log                   blog.Logger
+}
+
+var _ emailpb.ExporterServer = (*ExporterImpl)(nil)
+
+// NewExporterImpl initializes an ExporterImpl with the given client and
+// configuration. Both perDayLimit and maxConcurrentRequests should be
+// distributed proportionally among instances based on their share of the daily
+// request cap. For example, if the total daily limit is 50,000 and one instance
+// is assigned 40% (20,000 requests), it should also receive 40% of the max
+// concurrent requests (e.g., 2 out of 5). For more details, see:
+// https://developer.salesforce.com/docs/marketing/pardot/guide/overview.html?q=rate%20limits
+func NewExporterImpl(client PardotClient, perDayLimit float64, maxConcurrentRequests int, scope prometheus.Registerer, logger blog.Logger) *ExporterImpl {
+	limiter := rate.NewLimiter(rate.Limit(perDayLimit/86400.0), maxConcurrentRequests)
+
+	emailsHandledCounter := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "email_exporter_emails_handled",
+		Help: "Total number of emails handled by the email exporter",
+	})
+	scope.MustRegister(emailsHandledCounter)
+
+	impl := &ExporterImpl{
+		maxConcurrentRequests: maxConcurrentRequests,
+		limiter:               limiter,
+		toSend:                make([]string, 0, contactsQueueCap),
+		client:                client,
+		emailsHandledCounter:  emailsHandledCounter,
+		log:                   logger,
+	}
+	impl.wake = sync.NewCond(&impl.Mutex)
+
+	queueGauge := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "email_exporter_queue_length",
+		Help: "Current length of the email export queue",
+	}, func() float64 {
+		impl.Lock()
+		defer impl.Unlock()
+		return float64(len(impl.toSend))
+	})
+	scope.MustRegister(queueGauge)
+
+	return impl
+}
+
+// SendContacts enqueues the provided email addresses. If the queue cannot
+// accommodate the new emails, an ErrQueueFull is returned.
+func (impl *ExporterImpl) SendContacts(ctx context.Context, req *emailpb.SendContactsRequest) (*emptypb.Empty, error) {
+	if core.IsAnyNilOrZero(req, req.Emails) {
+		return nil, berrors.InternalServerError("Incomplete UpsertEmails request")
```

Comment:
```suggestion
		return nil, berrors.InternalServerError("Incomplete request")
```

Looks like the name of this method changed at some point? We don't need to include the name in the error message anyhow; that will wind up in the logs at the call site.